src/loaders/FileLoader.js

```diff
@@ -55,225 +55,191 @@ class FileLoader extends Loader {
 
-					for ( let i = 0, il = callbacks.length; i < il; i ++ ) {
+				switch ( this.responseType ) {
 
-						const callback = callbacks[ i ];
-						if ( callback.onLoad ) callback.onLoad( response );
-
-					}
-
-					scope.manager.itemEnd( url );
+					case 'arraybuffer':
 
-				} else {
+						return response.arrayBuffer();
 
-					for ( let i = 0, il = callbacks.length; i < il; i ++ ) {
+					case 'blob':
 
-						const callback = callbacks[ i ];
-						if ( callback.onError ) callback.onError( event );
+						return response.blob();
 
-					}
+					case 'document':
 
-					scope.manager.itemError( url );
-					scope.manager.itemEnd( url );
+						return response.text()
+							.then( text => {
 
-				}
+								const parser = new DOMParser();
+								return parser.parseFromString( text, this.mimeType );
 
-			}, false );
+							} );
 
-			request.addEventListener( 'progress', function ( event ) {
+					case 'json':
 
-				const callbacks = loading[ url ];
+						return response.json();
 
-				for ( let i = 0, il = callbacks.length; i < il; i ++ ) {
+					default:
 
-					const callback = callbacks[ i ];
-					if ( callback.onProgress ) callback.onProgress( event );
+						return response.text();
 
 				}
 
-			}, false );
+			} )
+			.then( data => {
 
-			request.addEventListener( 'error', function ( event ) {
+				// Add to cache only on HTTP success, so that we do not cache
+				// error response bodies as proper responses to requests.
+				Cache.add( url, data );
 
 				const callbacks = loading[ url ];
-
 				delete loading[ url ];
 
 				for ( let i = 0, il = callbacks.length; i < il; i ++ ) {
 
 					const callback = callbacks[ i ];
-					if ( callback.onError ) callback.onError( event );
+					if ( callback.onLoad ) callback.onLoad( data );
 
 				}
 
-				scope.manager.itemError( url );
 				scope.manager.itemEnd( url );
 
-			}, false );
+			} )
+			.catch( ( err ) => {
 
-			request.addEventListener( 'abort', function ( event ) {
+				// Abort errors and other errors are handled the same
 
 				const callbacks = loading[ url ];
-
 				delete loading[ url ];
 
 				for ( let i = 0, il = callbacks.length; i < il; i ++ ) {
 
 					const callback = callbacks[ i ];
-					if ( callback.onError ) callback.onError( event );
+					if ( callback.onError ) callback.onError( err );
 
 				}
 
 				scope.manager.itemError( url );
 				scope.manager.itemEnd( url );
 
-			}, false );
-
-			if ( this.responseType !== undefined ) request.responseType = this.responseType;
-			if ( this.withCredentials !== undefined ) request.withCredentials = this.withCredentials;
-
-			if ( request.overrideMimeType ) request.overrideMimeType( this.mimeType !== undefined ? this.mimeType : 'text/plain' );
-
-			for ( const header in this.requestHeader ) {
-
-				request.setRequestHeader( header, this.requestHeader[ header ] );
-
-			}
-
-			request.send( null );
-
-		}
+			} );
 
 		scope.manager.itemStart( url );
 
-		return request;
+		return;
```

Comment:
oops. will add `return request`  back in. The promise returned by the fetch is what everyone will need to attach their `.then()`'s to. Will add back once we confirm Safari can handle data URIs through fetch.

It still worked without it because it's been the `LoadingManager`'s job, not the `FileLoaders` to sort out.

Reply:
This is another thing I'm wondering what the apparent use cases for it are. As far as I know there aren't any loaders that use the return value of `FileLoader.load`. Perhaps we should just return nothing here, as well, and wait for people to complain?  @mrdoob?

Reply:
Haha, yes 😇

Reply:
haha lol. well I think I've figured out why I've been having trouble with `loadAsync()` up till now. It promises the return value of `load()`, which is the XMLRequest, but what did the XMLRequest load event return? nothing. It's the call-back which has the event/value. That's why when I've been trying to do `Promise.all([loadAsync(), ...])` it resolves early with `[undefined, ...]` but if inspected a small time later will have the correct value.

hmm because nobody has ever actually used the `return` we're probably in exactly the right spot to incrementally adopt `fetch` as it was intended. The call-backs can be called before the returned promises and the promises will simply be overwritten by the call-backs later on.

Folks can slowly migrate the loaders and then, if we all feel like it, take the plunge from `load( url, onLoad, onProgress, onError )` to `load( url )`.

Reply:
How will that compare to the current `loadAsync( url )`?

Reply:
hmm, if you'd like we could keep those kinds of changes to the returned value within `loadAsync()`. It'd be more or less copy the new `load()` but hey, that'd leave the old `load()` as is.

Reply:
Not changing behavior is good 👍